enhancements/20221228-various-storage-class-support.md

```diff
@@ -0,0 +1,114 @@
+
+### API changes
+
+No
+
+## Design
+
+### Implementation Overview
+
+PVC StorageClass will reference the `hostStorageClass`, which is from the StorageClass parameter or controller config hostStorageClass. The prioritize is StorageClass parameter > controller config. If w/o any definition, PVC would use the default StorageClass on the host cluster.
+
+We prefer to keep the default StorageClass, `harvester`, of harvester-CSI-driver. That would benefit the general cases because users do not need to change anything if they do not want extra StorageClass. If users need a new StorageClass for different purposes, they can create a new one.
+
+### StorageClass manifest Sample
+
+``` yaml
+  allowVolumeExpansion: true
+  apiVersion: storage.k8s.io/v1
+  kind: StorageClass
+  metadata:
+    name: <StorageClass Name>
+  provisioner: driver.harvesterhci.io
+  reclaimPolicy: Delete
+  volumeBindingMode: Immediate
+  parameters:
+    hostStorageClass: <StorageClass Name on the Host Harvester Cluster>
+```
+
+### Notible Variable
+
+- **hostStorageClass** - This is associated with the StorageClass Name on the host Harvester Cluster.
+
+### Flow Chart
+
+```mermaid
+flowchart TB
+    Start[Create Volume] -->|Some operations| GetVolParameters(Get Volume Parameters)
+  GetVolParameters --> GetHostStorageClassParam{Check `hostStorageClass` on Volume Parameters }
+  GetHostStorageClassParam -->|exist| D[Using `hostStorageClass` of volume parameters] --> CreateVolume
+  GetHostStorageClassParam -->|not exist| E[Using the `hostStorageClass` from configuration] --> CreateVolume
+  CreateVolume[Create volume and patch StorageClass with `hostStorageClass`]
+```
+
+### Test plan
+
+#### Verify the Default StorageClass behavior
+
+  1. Create the volume `vol001` with the default StorageClass on the Guest Cluster
+  2. Check the `vol001` is created with the Default StorageClass on the host Harvester Cluster.
+
+#### Verify the user-defined StorageClass behavior
+
+1. Create a new StorageClass `test` on the host Harvester Cluster.
+2. Create a new StorageClass using the following manifest on the Guest Cluster.
+
+    ``` yaml
+    allowVolumeExpansion: true
+    apiVersion: storage.k8s.io/v1
+    kind: StorageClass
+    metadata:
+      name: harvester-SC-test
+    provisioner: driver.harvesterhci.io
+    reclaimPolicy: Delete
+    volumeBindingMode: Immediate
+    parameters:
+      hostStorageClass: test
+    ```
+
+3. Create the volume `vol002` with the above StorageClass on the Guest Cluster.
+4. Check the `vol002` is created with the `test` StorageClass on the host Harvester Cluster.
+
+### Upgrade strategy
+
+No
+
+## Note [optional]
+
+### Rancher UI enhancement
+- The user needs to create StorageClass from manifest instead UI because the user could not select the provisioner `driver.harvesterhci.io` when creating StorageClass on UI.
```

Comment:
```suggestion
- The user needs to create a StorageClass from a manifest instead of the UI because the user can not select the provisioner `driver.harvesterhci.io` when creating a StorageClass through the UI.
```